core/federated/RTI/message_record/message_record.c

```diff
@@ -0,0 +1,170 @@
+/**
+ * @file message_record.c
+ * @author Soroush Bateni (soroush@berkeley.edu)
+ * @brief Record-keeping for in-transit messages.
+ * @version 0.1
+ * @date 2022-06-02
+ * 
+ * @copyright Copyright (c) 2022, The University of California at Berkeley.
+
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+***************/
+
+#include "message_record.h"
+#include <stdlib.h>
+
+/**
+ * @brief Initialize the in-transit message record queue.
+ * 
+ * @return in_transit_message_record_q 
+ */
+in_transit_message_record_q_t initialize_in_transit_message_q() {
+    in_transit_message_record_q_t queue = 
+        (in_transit_message_record_q_t)calloc(
+            1, 
+            sizeof(struct in_transit_message_record_q)
+        );
+    queue->main_queue = pqueue_init(
+        10, 
+        in_reverse_order, 
+        get_message_record_index,
+        get_message_record_position, 
+        set_message_record_position, 
+        tags_match, 
+        print_message_record
+    );
+
+    queue->transfer_queue = pqueue_init(
+        10, 
+        in_reverse_order, 
+        get_message_record_index,
+        get_message_record_position, 
+        set_message_record_position, 
+        tags_match, 
+        print_message_record
+    );
+
+    return queue;
+}
+
+/**
+ * @brief Free the memory occupied by the `queue`.
+ * 
+ * @param queue The queue to free.
+ */
+void free_in_transit_message_q(in_transit_message_record_q_t queue) {
+    pqueue_free(queue->main_queue);
+    pqueue_free(queue->transfer_queue);
+    free(queue);
```

Comment:
I think this free() is wrong because the argument is not a pointer. This will result in freeing random memory, or more likely in an error of some sort, assuming free() performs some checks.

Reply:
Should I perhaps rename `in_transit_message_record_q_t` to `in_transit_message_record_q_t_p` to make the fact that it is a pointer clearer? Or maybe we shouldn't hide pointers at all. This was just an arbitrary stylistic choice by me to abstract away the fact that `in_transit_message_record_q_t` is a pointer because the code outside message_record shouldn't care.

Reply:
I think it would be clearer to eliminate the following:
```
typedef struct in_transit_message_record_q* in_transit_message_record_q_t;
```
Instead of `in_transit_message_record_q_t`, just use `in_transit_message_record_q*`.

Reply:
Okay, done here: f05be58b838141ab2398b1f855fd8dd165a42e89